payments/src/lib.rs

```diff
@@ -0,0 +1,601 @@
+		PaymentDetail<T>,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn tasks)]
+	/// Store the list of tasks to be executed in the on_idle function
+	pub(super) type ScheduledTasks<T: Config> = StorageValue<_, ScheduledTaskList<T>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A new payment has been created
+		PaymentCreated {
+			from: T::AccountId,
+			asset: AssetIdOf<T>,
+			amount: BalanceOf<T>,
+			remark: Option<BoundedDataOf<T>>,
+		},
+		/// Payment amount released to the recipient
+		PaymentReleased { from: T::AccountId, to: T::AccountId },
+		/// Payment has been cancelled by the creator
+		PaymentCancelled { from: T::AccountId, to: T::AccountId },
+		/// A payment that NeedsReview has been resolved by Judge
+		PaymentResolved {
+			from: T::AccountId,
+			to: T::AccountId,
+			recipient_share: Percent,
+		},
+		/// the payment creator has created a refund request
+		PaymentCreatorRequestedRefund {
+			from: T::AccountId,
+			to: T::AccountId,
+			expiry: T::BlockNumber,
+		},
+		/// the refund request from creator was disputed by recipient
+		PaymentRefundDisputed { from: T::AccountId, to: T::AccountId },
+		/// Payment request was created by recipient
+		PaymentRequestCreated { from: T::AccountId, to: T::AccountId },
+		/// Payment request was completed by sender
+		PaymentRequestCompleted { from: T::AccountId, to: T::AccountId },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The selected payment does not exist
+		InvalidPayment,
+		/// The selected payment cannot be released
+		PaymentAlreadyReleased,
+		/// The selected payment already exists and is in process
+		PaymentAlreadyInProcess,
+		/// Action permitted only for whitelisted users
+		InvalidAction,
+		/// Payment is in review state and cannot be modified
+		PaymentNeedsReview,
+		/// Unexpeted math error
+		MathError,
+		/// Payment request has not been created
+		RefundNotRequested,
+		/// Dispute period has not passed
+		DisputePeriodNotPassed,
+		/// The automatic cancelation queue cannot accept
+		RefundQueueFull,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Hook that execute when there is leftover space in a block
+		/// This function will look for any pending scheduled tasks that can
+		/// be executed and will process them.
+		fn on_idle(now: T::BlockNumber, mut remaining_weight: Weight) -> Weight {
+			// reduce the weight used to read the task list
+			remaining_weight = remaining_weight.saturating_sub(T::WeightInfo::remove_task());
+
+			ScheduledTasks::<T>::mutate(|tasks| {
+				let mut task_list: Vec<_> = tasks
+					.clone()
+					.into_iter()
+					// leave out tasks in the future
+					.filter(|(_, ScheduledTask { when, task })| when <= &now && matches!(task, Task::Cancel))
+					.collect();
+
+				// order by oldest task to process
+				task_list.sort_by(|(_, t), (_, x)| x.when.partial_cmp(&t.when).unwrap());
```

Comment:
why use partial_cmp & unwrap instead of cmp?
also depends on number of expected tasks, we may want to change the data structure to have the tasks presorted

Reply:
The DS to use here is tricky, if we use a sorted BoundedVec<>, the `dispute_refund` extrinsic will get expensive, hence the use of map here.

Reply:
The DS to use here is tricky, if we use a sorted BoundedVec<>, the `dispute_refund` extrinsic will get expensive, hence the use of map here.